### ISSUE TITLE
Tests pass but with a warning to wrap in act when async operation occurs

### ISSUE THREAD

**Issue by antoniom** [testing-library/react-hooks-testing-library#742]
## Versions used
 - @testing-library/react: 12.1.0
 - @testing-library/react-hooks: 7.0.2
 - react: 17.0.2

## Sample code
```ts
// useFoo.ts
import React from 'react'

const getFoo = () => Promise.resolve('foo')

function useFoo(): string {
  const [value, setValue] = React.useState('')
  React.useEffect(() => {
    async function run() {
      const res = await getFoo()
      setValue(res)
    }

    run()
  }, [])

  return value
}

export default useFoo
```

```ts
// useFoo.test.ts
import { renderHook } from '@testing-library/react-hooks'
import useFoo from './useFoo'

describe('useFoo hook', () => {
  it('will return correct value', async () => {
    const { result, waitForNextUpdate } = renderHook(() => useFoo())

    await waitForNextUpdate()
    // also tried
    // await act(async () => {
    //    await waitForNextUpdate()
    // })

    expect(result.current).toEqual('foo')
  })
})
```

## What happens
Test passes with the following warning:
```
Warning: An update to TestComponent inside a test was not wrapped in act(...)
...
```
## What was expected
Test pass with no errors or warnings

Is it a bug? Am i missing something in the code?

**Comment by pol-sotos** [comment]
Hi, I am facing a similar issue. I am not be able to figure out which is the problem.
Thanks for your help.

Br

**Comment by mpeyper** [comment]
Hi @antoniom,

Is the hook you are testing different from what you've shared above?  I'm unable to replicate the warning with this one :

![image](https://user-images.githubusercontent.com/23029903/144652891-e5f0efda-beab-472a-961d-7835a318c12d.png).

Generally, this can occur when the condition to stop waiting is fulfilled, but the hook makes further updates after the test is completed.  A common case would be if there is further async work occurring, e.g.

```ts
React.useEffect(() => {
  async function run() {
    const res = await getFoo()
    setFoo(res)

    const res = await getBar()
    setBar(res)
  }

  run()
}, [])
```

The above effect would resolve the waiting promise when `setFoo` happens, run the assertion on `foo` and pass the test, then the update for `setBar` happens, triggering the `act` warning.

Try changing the test to something like this and see where the warning appears in relation to the numbers:

```ts
it('will return correct value', async () => {
  const { result, waitForNextUpdate } = renderHook(() => useFoo())

  console.log(1)
  await waitForNextUpdate()

  console.log(2)
  expect(result.current).toEqual('foo')

  console.log(3)
})
```

This should help narrow down where the rogue update is coming from.

**Comment by antoniom** [comment]
@mpeyper you are right, issue is not reproducible in a minimal environment that I have just set up!

But the exact same code does not work on my project's environment!
Is there any chance that some sort of configuration (in jest for example) to cause that problem? 
Unfortunately I don't have access on the project's code right now, so that I can check any weird jest configuration, but I will post any findings as soon as I check it.

**Comment by antoniom** [comment]
@mpeyper finally figured out why:

As testing-library's [documentation](https://testing-library.com/docs/react-testing-library/setup/#custom-render) suggests we created a custom render function and placed it in a `test-utils.tsx` module, exactly as described in the docs.
Then we changed the jest configuration, slightly different from what the [documentation](https://testing-library.com/docs/react-testing-library/setup/#configuring-jest-with-test-utils) suggests:
```js
// jest.config.js
{
  ...
  moduleNameMapper: {
    ...
    'test-utils': '<rootDir>/src/utils/test-utils.tsx'
  }
}
```
instead of 
```js
{
  ...
  moduleDirectories: [
    'node_modules',
   '<rootDir>/src/utils', // 
    __dirname, 
  ],
```

I suppose that test-utils inside moduleNameMapper acts as a regex and tries to override that `react-dom/test-utils` used inside `react-hooks-testing-library`.

Everything seems to be working fine now,

Thanks for taking the time to investigate the issue!
